File: src/lexer/lexer.go
package lexer

import (
	"bytes"
	"unicode"
	"vesti/src/location"
	"vesti/src/newlineHandler"
	"vesti/src/token"
)

type LexToken struct {
	Token token.Token
	Span  location.Span
}

type Lexer struct {
	source          *newlineHandler.NewlineHandler
	chr0            rune
	chr1            rune
	chr2            rune
	currentLocation *location.Location
	mathStarted     bool
}

func New(input string) *Lexer {
	lex := &Lexer{
		source: newlineHandler.New(input),
		chr0:   0, chr1: 0, chr2: 0,
		currentLocation: location.New(),
		mathStarted:     false,
	}

	lex.nextChar()
	lex.nextChar()
	lex.nextChar()
	lex.currentLocation.ResetLocation()

	return lex
}

func (l *Lexer) nextChar() {
	if l.chr0 == '\n' {
		l.currentLocation.MoveNextLine()
	} else {
		l.currentLocation.MoveRight()
	}

	l.chr0 = l.chr1
	l.chr1 = l.chr2
	l.chr2 = l.source.Next()
}

func (l *Lexer) tokenize(lexTok *LexToken, tok token.TokenType, lit string) {
	l.nextChar()
	lexTok.Token = token.New(tok, lit)
	lexTok.Span.End = *l.currentLocation
}

func (l *Lexer) TakeTok() *LexToken {
	var output = LexToken{Span: location.Span{Start: *l.currentLocation}}

	switch l.chr0 {
	case 0:
		l.tokenize(&output, token.EOF, "")
	case ' ':
		l.tokenize(&output, token.Space, " ")
	case '\t':
		l.tokenize(&output, token.Tab, "\t")
	case '\n':
		l.tokenize(&output, token.Newline, "\n")
	case '+':
		if l.chr1 == '#' && l.chr2 == '#' {
			l.nextChar()
			l.nextChar()
			l.tokenize(&output, token.ObeyNewlineBeforeDocEnd, "")
		} else {
			l.tokenize(&output, token.Plus, "+")
		}
	case '-':
		if unicode.IsDigit(l.chr1) || l.chr1 == '.' {
			l.lexNumber(&output)
		} else {
			l.tokenize(&output, token.Minus, "-")
		}
	case '*':
		l.tokenize(&output, token.Star, "*")
	case '/':
		l.tokenize(&output, token.Slash, "/")
	case '=':
		l.tokenize(&output, token.Equal, "=")
	case '<':
		if l.chr1 == '=' {
			l.nextChar()
			l.tokenize(&output, token.LessEq, "<=")
		} else {
			l.tokenize(&output, token.Less, "<")
		}
	case '>':
		if l.chr1 == '=' {
			l.nextChar()
			l.tokenize(&output, token.GreatEq, ">=")
		} else {
			l.tokenize(&output, token.Great, ">")
		}
	case '!':
		l.tokenize(&output, token.Bang, "!")
	case '?':
		l.tokenize(&output, token.Question, "?")
	case '@':
		l.tokenize(&output, token.At, "@")
	case '%':
		l.tokenize(&output, token.Percent, "%")
	case '^':
		l.tokenize(&output, token.Superscript, "^")
	case '_':
		l.tokenize(&output, token.Subscript, "_")
	case '&':
		l.tokenize(&output, token.Ampersand, "&")
	case ';':
		l.tokenize(&output, token.Semicolon, ";")
	case ':':
		l.tokenize(&output, token.Colon, ":")
	case '\'':
		l.tokenize(&output, token.Quote, "'")
	case '`':
		l.tokenize(&output, token.Quote2, "`")
	case '"':
		l.tokenize(&output, token.Doublequote, "\"")
	case '|':
		l.tokenize(&output, token.Vert, "|")
	case '.':
		if unicode.IsDigit(l.chr1) {
			l.lexNumber(&output)
		} else {
			l.tokenize(&output, token.Period, ".")
		}
	case ',':
		l.tokenize(&output, token.Comma, ",")
	case '~':
		l.tokenize(&output, token.Tilde, "~")
	case '(':
		l.tokenize(&output, token.Lparen, "(")
	case ')':
		l.tokenize(&output, token.Rparen, ")")
	case '{':
		l.tokenize(&output, token.Lbrace, "{")
	case '}':
		l.tokenize(&output, token.Rbrace, "}")
	case '[':
		l.tokenize(&output, token.Lsqbrace, "[")
	case ']':
		l.tokenize(&output, token.Rsqbrace, "]")
	case '$':
		l.lexMathToken(&output)
	case '#':
		l.lexSharp(&output)
	case '\\':
		l.lexBackslash(&output)
	default:
		switch {
		case unicode.IsLetter(l.chr0):
			l.lexMainString(&output)
		case unicode.IsDigit(l.chr0):
			l.lexNumber(&output)
		default:
			l.tokenize(&output, token.ILLEGAL, "")
		}
	}

	return &output
}

func (l *Lexer) lexMainString(tok *LexToken) {
	var literal bytes.Buffer

	for l.chr0 != 0 {
		if !unicode.IsLetter(l.chr0) && !unicode.IsDigit(l.chr0) {
			break
		}
		literal.WriteRune(l.chr0)
		l.nextChar()
	}

	tok.Token.Literal = literal.String()
	tok.Token.Type = token.Lookup(tok.Token.Literal)
	tok.Span.End = *l.currentLocation
}

func (l *Lexer) lexNumber(tok *LexToken) {
	var literal bytes.Buffer
	var tokType token.TokenType

	if l.chr0 == '-' {
		literal.WriteRune('-')
		l.nextChar()
	}

	if l.chr0 != '.' {
		for unicode.IsDigit(l.chr0) {
			literal.WriteRune(l.chr0)
			l.nextChar()
		}
	}

	if l.chr0 == '.' {
		literal.WriteRune('.')
		l.nextChar()
		tokType = token.Float
	} else {
		tokType = token.Integer
	}

	for unicode.IsDigit(l.chr0) {
		literal.WriteRune(l.chr0)
		l.nextChar()
	}

	tok.Token = token.New(tokType, literal.String())
	tok.Span.End = *l.currentLocation
}

func (l *Lexer) lexMathToken(tok *LexToken) {
	if l.mathStarted {
		l.mathStarted = false
		if l.chr1 == '$' {
			l.nextChar()
			l.tokenize(tok, token.InlineMathEnd, "\\]")
		} else {
			l.tokenize(tok, token.TextMathEnd, "$")
		}
	} else {
		l.mathStarted = true
		if l.chr1 == '$' {
			l.nextChar()
			l.tokenize(tok, token.InlineMathStart, "\\[")
		} else {
			l.tokenize(tok, token.TextMathStart, "$")
		}
	}
}

func (l *Lexer) lexSharp(tok *LexToken) {
	switch l.chr1 {
	case '*':
		l.nextChar()
		l.nextChar()
		for l.chr0 != '*' || l.chr1 != '#' {
			if l.chr0 == 0 {
				break
			}
			l.nextChar()
		}
		l.nextChar()
		l.nextChar()
		start := tok.Span.Start
		*tok = *l.TakeTok()
		tok.Span.Start = start
	case '!':
		var literal bytes.Buffer

		l.nextChar()
		l.nextChar()
		for l.chr0 != '!' || l.chr1 != '#' {
			if l.chr0 == 0 {
				break
			}
			literal.WriteRune(l.chr0)
			l.nextChar()
		}
		l.nextChar()
		l.nextChar()
		tok.Token = token.New(token.RawLatex, literal.String())
		tok.Span.End = *l.currentLocation
	case '@':
		l.nextChar()
		l.tokenize(tok, token.Newline2, "\n")
	default:
		switch {
		case l.chr1 == '#' && l.chr2 == '-':
			var literal bytes.Buffer

			l.nextChar()
			l.nextChar()
			l.nextChar()
			for l.chr0 != '-' || l.chr1 != '#' || l.chr2 != '#' {
				if l.chr0 == 0 {
					break
				}
				literal.WriteRune(l.chr0)
				l.nextChar()
			}
			l.nextChar()
			l.nextChar()
			l.nextChar()
			tok.Token = token.New(token.RawLatex, literal.String())
			tok.Span.End = *l.currentLocation
		case l.chr1 == '#' && l.chr2 == '+':
			l.nextChar()
			l.nextChar()
			l.tokenize(tok, token.ObeyNewlineBeforeDocStart, "")
		default:
			for l.chr0 != '\n' {
				if l.chr0 == 0 {
					break
				}
				l.nextChar()
			}
			start := tok.Span.Start
			*tok = *l.TakeTok()
			tok.Span.Start = start
		}
	}
}

func (l *Lexer) lexBackslash(tok *LexToken) {
	switch l.chr1 {
	// TODO: I can't find much more elegant token to represent latex define function parameter
	case '?':
		l.nextChar()
		l.tokenize(tok, token.FntParam, "#")
	case '#':
		l.nextChar()
		l.tokenize(tok, token.Sharp, "\\#")
	case '$':
		l.nextChar()
		l.tokenize(tok, token.Dollar, "\\$")
	case ',':
		l.nextChar()
		if l.mathStarted {
			l.tokenize(tok, token.MathSmallSpace, "\\,")
		} else {
			l.tokenize(tok, token.Comma, ",")
		}
	case ';':
		l.nextChar()
		l.tokenize(tok, token.ArgSpliter, "")
	case '(':
		l.nextChar()
		l.tokenize(tok, token.TextMathStart, "$")
	case ')':
		l.nextChar()
		l.tokenize(tok, token.TextMathEnd, "$")
	case '{':
		l.nextChar()
		l.tokenize(tok, token.MathLbrace, "\\{")
	case '}':
		l.nextChar()
		l.tokenize(tok, token.MathRbrace, "\\}")
	case '[':
		l.nextChar()
		l.tokenize(tok, token.InlineMathStart, "\\[")
	case ']':
		l.nextChar()
		l.tokenize(tok, token.InlineMathEnd, "\\]")
	case ' ':
		l.nextChar()
		if l.mathStarted {
			l.tokenize(tok, token.MathLargeSpace, "\\;")
		} else {
			l.tokenize(tok, token.Space2, "\\ ")
		}
	case '\\':
		l.nextChar()
		l.tokenize(tok, token.BackSlash, "\\\\")
	default:
		if unicode.IsLetter(l.chr1) || l.chr1 == '@' {
			var literal bytes.Buffer

			l.nextChar()
			for unicode.IsLetter(l.chr0) || l.chr0 == '@' {
				tmpTok := l.TakeTok()
				// There are only two tokens, EOF and ILLEGAL, that are larger than 0xE000
				if tmpTok.Token.Type > 0xE000 {
					break
				}
				literal.WriteString(tmpTok.Token.Literal)
			}
			tok.Token = token.New(token.LatexFunction, literal.String())
			tok.Span.End = *l.currentLocation
		} else {
			l.tokenize(tok, token.ILLEGAL, "")
		}
	}
}

File: src/Lexer.zig
const std = @import("std");
const ziglyph = @import("ziglyph");
const location = @import("location.zig");

const Token = @import("token.zig").Token;
const Type = @import("token.zig").Type;

fn isLatexFunctionIdent(chr: u21) bool {
    return chr == '@' or ziglyph.isAlphabetic(chr);
}

fn isOtherUnicode(chr: u21) bool {
    return ziglyph.isPrint(chr) and chr > 0xff and !ziglyph.isAlphaNum(chr);
}

pub const Lexer = @This();
const Self = @This();

// Lexer Fields
source: []const u8,
index: usize = 0,
location: location.Location,
math_start: bool = false,
// END Fields

const Chars = struct {
    chr0: ?u21,
    chr1: ?u21,
};

fn defaultLit(self: *const Self, start: usize) []const u8 {
    return self.source[start..self.index];
}

fn getChars(self: *Self, codepoint_len: *u3) Chars {
    var output: Chars = .{ .chr0 = null, .chr1 = null };
    if (self.index >= self.source.len) {
        return output;
    }
    codepoint_len.* = std.unicode.utf8ByteSequenceLength(self.source[self.index]) catch unreachable;
    output.chr0 = std.unicode.utf8Decode(self.source[self.index .. self.index + codepoint_len.*]) catch unreachable;
    if (self.index + codepoint_len.* >= self.source.len) {
        return output;
    }
    const next_codepoint_len = std.unicode.utf8ByteSequenceLength(self.source[self.index + codepoint_len.*]) catch unreachable;
    output.chr1 = std.unicode.utf8Decode(self.source[self.index + codepoint_len.* .. self.index + codepoint_len.* + next_codepoint_len]) catch unreachable;

    return output;
}

pub fn new(source: []const u8) Self {
    return .{ .source = source, .location = .{ .row = 1, .column = 1 } };
}

pub fn next(self: *Self) Token {
    var state: enum {
        start,
        minus,
        equal,
        slash,
        bang,
        less,
        great,
        percent,
        dollar,
        at,
        backslash,
        text,
        zero,
        integers,
        floats,
        etc_unicode,
        latex_function,
        line_comment,
        multi_line_comment,
        multi_line_comment_star,
        multi_line_comment_end,
        verbatim,
        verbatim_end,
    } = .start;

    var start_index = self.index;
    var start_location = self.location;
    var toktype: Type = .eof;
    var literal: []const u8 = "";

    var codepoint_len: u3 = undefined;
    while (self.index < self.source.len) : (self.index += codepoint_len) {
        const chars = self.getChars(&codepoint_len);
        const chr0 = chars.chr0.?;
        switch (state) {
            .start => switch (chr0) {
                ' ' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .space;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '\t' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .tab;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '\n' => {
                    self.index += 1;
                    self.location.newLine();
                    toktype = .newline;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '+' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .plus;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '*' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .star;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '.' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .period;
                    literal = self.defaultLit(start_index);
                    break;
                },
                ',' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .comma;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '?' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .question;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '&' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .ampersand;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '^' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .superscript;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '_' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .subscript;
                    literal = self.defaultLit(start_index);
                    break;
                },
                ';' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .semicolon;
                    literal = self.defaultLit(start_index);
                    break;
                },
                ':' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .colon;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '`' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .left_quote;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '\'' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .right_quote;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '"' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .double_quote;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '|' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .vert;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '~' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .tilde;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '(' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .left_paren;
                    literal = self.defaultLit(start_index);
                    break;
                },
                ')' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .right_paren;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '{' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .left_brace;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '}' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .right_brace;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '[' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .left_square_brace;
                    literal = self.defaultLit(start_index);
                    break;
                },
                ']' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .right_square_brace;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '#' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .function_param;
                    literal = self.defaultLit(start_index);
                    break;
                },
                '>' => {
                    self.location.moveRight(chr0);
                    state = .great;
                },
                '/' => {
                    self.location.moveRight(chr0);
                    state = .slash;
                },
                '!' => {
                    self.location.moveRight(chr0);
                    state = .bang;
                },
                '-' => {
                    self.location.moveRight(chr0);
                    state = .minus;
                },
                '=' => {
                    self.location.moveRight(chr0);
                    state = .equal;
                },
                '<' => {
                    self.location.moveRight(chr0);
                    state = .less;
                },
                '@' => {
                    self.location.moveRight(chr0);
                    state = .at;
                },
                '%' => {
                    self.location.moveRight(chr0);
                    state = .percent;
                },
                '$' => {
                    self.location.moveRight(chr0);
                    state = .dollar;
                },
                '\\' => {
                    self.location.moveRight(chr0);
                    state = .backslash;
                },
                '0' => {
                    self.location.moveRight(chr0);
                    state = .zero;
                },
                '1'...'9' => {
                    self.location.moveRight(chr0);
                    state = .integers;
                },
                else => if (ziglyph.isAlphaNum(chr0)) {
                    self.location.moveRight(chr0);
                    state = .text;
                } else if (isOtherUnicode(chr0)) {
                    self.location.moveRight(chr0);
                    state = .etc_unicode;
                } else {
                    self.index += 1;
                    toktype = .illegal;
                    literal = self.defaultLit(start_index);
                    break;
                },
            },
            .slash => switch (chr0) {
                '=' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .not_equal;
                    literal = "\\neq ";
                    break;
                },
                else => {
                    toktype = .slash;
                    literal = self.defaultLit(start_index);
                    break;
                },
            },
            .bang => switch (chr0) {
                '=' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .not_equal;
                    literal = "\\neq ";
                    break;
                },
                else => {
                    toktype = .bang;
                    literal = self.defaultLit(start_index);
                    break;
                },
            },
            .minus => switch (chr0) {
                '>' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .right_arrow;
                    literal = "\\to ";
                    break;
                },
                '0' => {
                    self.location.moveRight(chr0);
                    state = .zero;
                },
                '1'...'9' => {
                    self.location.moveRight(chr0);
                    state = .integers;
                },
                else => {
                    toktype = .minus;
                    literal = self.defaultLit(start_index);
                    break;
                },
            },
            .equal => switch (chr0) {
                '>' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .double_right_arrow;
                    literal = "\\Rightarrow ";
                    break;
                },
                else => {
                    toktype = .equal;
                    literal = self.defaultLit(start_index);
                    break;
                },
            },
            .great => switch (chr0) {
                '=' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .great_equal;
                    literal = "\\geq ";
                    break;
                },
                else => {
                    toktype = .great;
                    literal = self.defaultLit(start_index);
                    break;
                },
            },
            .less => switch (chr0) {
                '-' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .left_arrow;
                    literal = "\\leftarrow ";
                    break;
                },
                '=' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .less_equal;
                    literal = "\\leq ";
                    break;
                },
                else => {
                    toktype = .less;
                    literal = self.defaultLit(start_index);
                    break;
                },
            },
            .at => switch (chr0) {
                '!' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .at;
                    literal = "@";
                    break;
                },
                else => {
                    toktype = .argument_splitter;
                    literal = "@";
                    break;
                },
            },
            .percent => switch (chr0) {
                '!' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .latex_comment;
                    literal = "%";
                    break;
                },
                '-' => {
                    self.location.moveRight(chr0);
                    state = .verbatim;
                },
                '*' => {
                    self.location.moveRight(chr0);
                    state = .multi_line_comment;
                },
                '\n' => {
                    self.index += 1;
                    self.location.newLine();
                    toktype = .newline;
                    literal = "\n";
                    break;
                },
                else => {
                    self.location.moveRight(chr0);
                    state = .line_comment;
                },
            },
            .dollar => switch (chr0) {
                '!' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .raw_dollar;
                    literal = "$";
                    break;
                },
                '$' => if (self.math_start) {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .display_math_end;
                    literal = "\\]";
                    self.math_start = false;
                    break;
                } else {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .display_math_start;
                    literal = "\\[";
                    self.math_start = true;
                    break;
                },
                else => if (self.math_start) {
                    toktype = .text_math_end;
                    literal = "$";
                    self.math_start = false;
                    break;
                } else {
                    toktype = .text_math_start;
                    literal = "$";
                    self.math_start = true;
                    break;
                },
            },
            .backslash => switch (chr0) {
                '#' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .sharp;
                    literal = "\\#";
                    break;
                },
                '%' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .percent;
                    literal = "\\%";
                    break;
                },
                '$' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .dollar;
                    literal = "\\$";
                    break;
                },
                '(' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    self.math_start = true;
                    toktype = .text_math_start;
                    literal = "$";
                    break;
                },
                ')' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    self.math_start = false;
                    toktype = .text_math_end;
                    literal = "$";
                    break;
                },
                '[' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    self.math_start = true;
                    toktype = .display_math_start;
                    literal = "\\[";
                    break;
                },
                ']' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    self.math_start = false;
                    toktype = .display_math_end;
                    literal = "\\]";
                    break;
                },
                '{' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .math_left_brace;
                    literal = "\\{";
                    break;
                },
                '}' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .math_right_brace;
                    literal = "\\}";
                    break;
                },
                ' ' => if (self.math_start) {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .math_large_space;
                    literal = "\\;";
                    break;
                } else {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .force_space;
                    literal = "\\ ";
                    break;
                },
                '\\' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .backslash;
                    literal = "\\\\";
                    break;
                },
                else => if (isLatexFunctionIdent(chr0)) {
                    self.location.moveRight(chr0);
                    state = .latex_function;
                } else {
                    toktype = .short_backslash;
                    literal = "\\";
                    break;
                },
            },
            .text => if (!ziglyph.isAlphaNum(chr0)) {
                literal = self.defaultLit(start_index);
                toktype = Type.keywords.get(literal) orelse .text;
                break;
            } else {
                self.location.moveRight(chr0);
            },
            .zero => switch (chr0) {
                '0'...'9' => {
                    self.location.moveRight(chr0);
                    state = .text;
                },
                '.' => {
                    if (chars.chr1) |chr1| {
                        switch (chr1) {
                            '0'...'9' => {
                                self.location.moveRight(chr0);
                                state = .floats;
                            },
                            else => {
                                toktype = .integer;
                                literal = self.defaultLit(start_index);
                                break;
                            },
                        }
                    } else {
                        toktype = .integer;
                        literal = self.defaultLit(start_index);
                        break;
                    }
                },
                else => {
                    toktype = .integer;
                    literal = self.defaultLit(start_index);
                    break;
                },
            },
            .integers => {
                switch (chr0) {
                    '0'...'9' => self.location.moveRight(chr0),
                    '.' => {
                        if (chars.chr1) |chr1| {
                            switch (chr1) {
                                '0'...'9' => {
                                    self.location.moveRight(chr0);
                                    state = .floats;
                                },
                                else => {
                                    toktype = .integer;
                                    literal = self.defaultLit(start_index);
                                    break;
                                },
                            }
                        } else {
                            toktype = .integer;
                            literal = self.defaultLit(start_index);
                            break;
                        }
                    },
                    else => {
                        toktype = .integer;
                        literal = self.defaultLit(start_index);
                        break;
                    },
                }
            },
            .floats => {
                switch (chr0) {
                    '0'...'9' => self.location.moveRight(chr0),
                    else => {
                        toktype = .float;
                        literal = self.defaultLit(start_index);
                        break;
                    },
                }
            },
            .etc_unicode => if (!isOtherUnicode(chr0)) {
                toktype = .other_unicode_char;
                literal = self.defaultLit(start_index);
                break;
            },
            .latex_function => if (!isLatexFunctionIdent(chr0)) {
                toktype = .latex_function;
                literal = self.defaultLit(start_index);
                break;
            } else {
                self.location.moveRight(chr0);
            },
            .line_comment => switch (chr0) {
                '\n' => {
                    self.index += 1;
                    self.location.newLine();
                    toktype = .newline;
                    literal = "\n";
                    break;
                },
                '\r' => unreachable,
                else => self.location.moveRight(chr0),
            },
            .multi_line_comment => switch (chr0) {
                '*' => {
                    self.location.moveRight(chr0);
                    state = .multi_line_comment_star;
                },
                '\n' => self.location.newLine(),
                else => self.location.moveRight(chr0),
            },
            .multi_line_comment_star => switch (chr0) {
                '%' => {
                    self.location.moveRight(chr0);
                    state = .multi_line_comment_end;
                },
                '*' => self.location.moveRight(chr0),
                '\n' => {
                    self.location.newLine();
                    state = .multi_line_comment;
                },
                else => {
                    self.location.moveRight(chr0);
                    state = .multi_line_comment;
                },
            },
            .multi_line_comment_end => switch (chr0) {
                '\n' => {
                    self.index += 1;
                    self.location.newLine();
                    toktype = .newline;
                    literal = "\n";
                    break;
                },
                '\r' => unreachable,
                ' ' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .space;
                    literal = " ";
                    break;
                },
                '\t' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .tab;
                    literal = "\t";
                    break;
                },
                else => {
                    toktype = .nop;
                    literal = "";
                    break;
                },
            },
            .verbatim => switch (chr0) {
                '-' => {
                    self.location.moveRight(chr0);
                    state = .verbatim_end;
                },
                '\n' => self.location.newLine(),
                else => self.location.moveRight(chr0),
            },
            .verbatim_end => switch (chr0) {
                '%' => {
                    self.index += 1;
                    self.location.moveRight(chr0);
                    toktype = .raw_latex;
                    literal = self.source[start_index + 2 .. self.index - 2];
                    break;
                },
                '-' => self.location.moveRight(chr0),
                '\n' => {
                    self.location.newLine();
                    state = .verbatim;
                },
                else => {
                    self.location.moveRight(chr0);
                    state = .verbatim;
                },
            },
        }
    } else if (self.index >= self.source.len) {
        switch (state) {
            .start, .line_comment, .percent => {},
            .minus => {
                toktype = .minus;
                literal = "-";
            },
            .equal => {
                toktype = .equal;
                literal = "=";
            },
            .slash => {
                toktype = .slash;
                literal = "/";
            },
            .bang => {
                toktype = .bang;
                literal = "!";
            },
            .less => {
                toktype = .less;
                literal = "<";
            },
            .great => {
                toktype = .great;
                literal = ">";
            },
            .at => {
                toktype = .argument_splitter;
                literal = "";
            },
            .dollar => if (self.math_start) {
                toktype = .text_math_end;
                literal = "$";
            } else {
                toktype = .text_math_start;
                literal = "$";
            },
            .text => {
                literal = self.defaultLit(start_index);
                toktype = Type.keywords.get(literal) orelse .text;
            },
            .zero => {
                toktype = .integer;
                literal = self.defaultLit(start_index);
            },
            .integers => {
                toktype = .integer;
                literal = self.defaultLit(start_index);
            },
            .floats => {
                toktype = .float;
                literal = self.defaultLit(start_index);
            },
            .etc_unicode => {
                toktype = .other_unicode_char;
                literal = self.defaultLit(start_index);
            },
            .latex_function => {
                toktype = .latex_function;
                literal = self.defaultLit(start_index);
            },
            .multi_line_comment_end => {
                toktype = .newline;
                literal = "\n";
            },
            .backslash,
            .multi_line_comment,
            .multi_line_comment_star,
            .verbatim,
            .verbatim_end,
            => toktype = .illegal,
        }
    }

    return .{
        .toktype = toktype,
        .literal = literal,
        .span = .{ .start = start_location, .end = self.location },
    };
}
